### ISSUE TITLE
Support for squashing migrations for trunk based development workflows

### ISSUE THREAD

**Issue by jasonkuhrt** [prisma/prisma#5714]
Our team is thinking about how to adopt an ideal (but simple) migration workflow.

We already practice [trunk based development](https://trunkbaseddevelopment.com/) on our application.

This means we branch of trunk, make changes, create a pull-request, iterate (multiple commits) with a preview deploy for that branch, then once done, squash-merge the work into trunk where it appears as a single commit of work. The merge into trunk kicks off a production deployment (with automated tests of course).

The way we see adopting Prisma into this workflow right now:

1. Commits should have max 1 migration. This is for simplicity. We can't think why a commit should have multiple migrations since a commit is an atomic deployable unit. E.g. there will never be a deployment of "half of a commit".

2. On branches, eventually, migrations should be deployed to the preview database by CI runs triggered by the git push. But for now we're ok doing manual migration deploys via the CLI right before pushing the commit.

3. A branch has many commits but as mentioned before we want trunk to represent the work as one commit.  This can lead to a situation where the resulting 1 trunk commit will have more than 1 migration, violating our number 1 workflow rule. Therefore squash-merging a PR we will locally delete all the migrations introduced by the branch and then re-create just one migration. We'll commit+push that to the branch and then squash merge it (if we're being pick, probably way for the branch preview deploy yet again one last time etc. but in practice we'll probably skip this).

This issue is a feature request for step 3 above.

It would be great if instead of doing this manually:

> Therefore squash-merging a PR we will locally delete all the migrations introduced by the branch and then re-create just one migration.

We could have tooling to do this. Literally that initially seems to translate to:

```
prisma migrate squash --from <name> --to <name>
``` 

But actually we would love to introduce a PR bot that adds the commits to the branch for us, thus keeping the PR continually in a ready-for-trunk-merge state (respective to our number 1 rule above).

That would require us to:

1. Find out the migrations that have been introduced on this branch
2. Tell Prisma to squash merge them all

Thus this form would be advantageous to us:

```
prisma migrate squash --from <name>
``` 

This way we would just look for the oldest migration introduced in the branch, then let Prisma infer to squash up and including the latest migration.

I don't know if it would be appropriate for Prisma to including support for discovering the migrations introduced by a git branch over its base... This would involve getting cozy with `git` internally which is probably going to far in scope here, e.g.:

```
prisma migrate list --from-current-branch
``` 

And thus, e.g.:

```
prisma migrate squash --from $(prisma migrate list --json --from-current-branch | jq '[0]')
``` 

That said, as a user, having `prisma migrate list --from-current-branch` would greatly lower the bar for me to implement my automation. I would avoid having to:

- Think about/Get familiar with how migrations are ordered on disc

- Script/learn git, especially the nuances of branches, finding their parents, then diffing from state of their parents, blah blah blah

- handling edge cases (like when a branch introduces prisma, no migrations history to diff on parent branch)

Hmm... I guess, for us, while squash is a necessary ingredient here, it _alone_ doesn't really save us effort/time/thinking than our current manual approach:

> Therefore squash-merging a PR we will locally delete all the migrations introduced by the branch and then re-create just one migration.

Its only once we get the PR CI automation going that things become 💯 

Finally, a point raised by @albertoperdomo on slack, the challenges introduced by human edits to migration files for squash logic.

My assumption right now as a user is that if I go manually edit a migration I am giving up being able to squash it later.

**Comment by Jolg42** [comment]
Manual squashing is possible. The way I would do this, assuming you want to squash the migrations _at the end of your migrations history_, and that the migrations haven't been applied outside of your dev environment, is the following:

- Delete the relevant migration folders
- Run `migrate dev` again. It will ask you to reset the database, then create the squashed migration
- In case one of the squashed migrations was applied outside of your dev machine, you will need to baseline (`migrate resolve --applied <name-of-the-squashed-migration>` there.

Note that this will regenerate the migration, so if you manually changed the SQL, these changes will be gone. You could also manually squash (move all the SQL in a single migration) if this is what you need.

It's not the most readable resource, but we do also have [internal tests](https://github.com/prisma/prisma-engines/blob/53dc252c7ac8d195ff6a341612b538c91de5982e/migration-engine/migration-engine-tests/tests/migrations/squashing_tests.rs) that make sure this process of squashing works. It's definitely something we can make more ergonomic in the future, it's just a matter of making plans and executing on them :)

_Originally posted by @tomhoule in https://github.com/prisma/prisma/issues/5226#issuecomment-852804241_

**Comment by leumasme** [comment]
With the addition of `prisma db push` (which syncs Database schema to Prisma schema without creating a migration), squashing migrations wouldnt be as nessecary anymore.
The problem with this is that, when then later creating the migration with `prisma migrate dev`, the database will be reset as prisma detects a desync between the current database state and the migration history - this is nonsense in my eyes, and should be changed as its quite the inconvenience.
If this is fixed, squashing migrations could be replaced by not actually creating migrations until the final schema is archieved.

**Comment by dejavu1987** [comment]
In my case and anybody elses I guess, I want the exact opposite of what prisma db push does.
Scenario is:
I have applied several incremental migrations to my dev env already as I keep updating the schema.
I would want to delete the migrations, delete the migration records in the db.
At this point, what I am left with, is something `prisma db push` would have done.
But from here I would want to create a new migration, (without loosing data)
- with SQL for all the schema changes that had happened after last migration,
- entry in migration db table saying the new migration has been applied as if I did `resolve applied`.

I am using a workaround which is not very convenient,
- I backup my db
- Delete migrations
- Create migrations (wipes the db)
- restore db with the backup I created initially except the migration table.

In the end the db has all the changes to schema already anyway I just needed that migration and the db entry.

**Comment by do4gr** [comment]
Hey @jasonkuhrt, 

can you test whether this workflow with the new migrate diff command would work for your case: 
1.  git reset local migration history folder to migration history from main
2. create an empty migration (folder +file) in ./migrations -> save the file path (This could potentially be a new migrate command)
3. npx prisma migrate diff --preview-feature --from-migrations ./migrations --to-schema-datamodel ./schema.prisma --script > filepath

This would give you one migration that includes all the changes necessary to go from the state of migrations on main to your current datamodel on your branch.

**Comment by huksley** [comment]
I think that instead of permanently squashing migrations into one (and removing existing) it is better to make a cummulative migration which includes some previous ones. So migration folder will contain both unsquashed and cummulative and prisma can decide should it apply individually or not.

That way databases in the middle of migration group can be also updated. But you could delete individual migrations without affecting group ones.

So workflow would be something.

1. prisma migrate dev
2. prisma migrate group => server/prisma/migrations/20240918130839_cumulative
3. apply iteratively on all database instances
4. `remove manually everything prior server/prisma/migrations/20240918130839_cumulative`

**Comment by GregB-LADB** [comment]
I just came here to make a suggestion to add a `prisma squash` command as well.  I like this idea especially with a range of migrations to squash into one.  I took over a project that's been going on for a few years and the migration list of the schema is huge, with no live data yet.  It would be nice to make today's version of the schema the baseline.  I could certainly drop the db and start over from the current schema, but there is some test data that would be nice to keep without also exporting and reimporting that. 

So a built in squash command would be perfect instead of the multiple manual steps it currently takes.

**Comment by 0tii** [comment]
honestly I don't see a practical use for squashing like that for actual enterprise-scale projects, there has to be a method to squash WITH manual edits, other than writing custom concatenation scripts.

Our project has around 150 migrations and of those, easily 100 have custom triggers, checks, partial unique indices and whatever else that can't be modeled in the prisma schema.

Not being able to cleanly squash those without custom script work is a problem.
